### DISCUSSION TITLE
Inconsistent Initial Value Retrieval with atomWithStorage when getOnInit is True

### DISCUSSION THREAD

**Discussion by chiwon1** [pmndrs/jotai#2305] [Bug report] [answered]
I am encountering an issue with the atomWithStorage function in Jotai, where it fails to correctly retrieve the initial value from localStorage despite the getOnInit option being set to true. This problem occurs even though the atom should immediately reflect the value stored in localStorage upon initialization. Notably, when I use window.localStorage directly in my component, it behaves as expected, suggesting the issue is specific to Jotai's implementation.

Environment
React Version: 18.2.0
Jotai Version: 2.6.0
Other relevant libraries/frameworks: Create React App
Package.json contents:
```json
{
  "name": "test",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^5.17.0",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^13.5.0",
    "jotai": "^2.6.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-scripts": "5.0.1",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

Reproduction
Here's the code snippet that demonstrates the issue:

```jsx
import "./App.css";
import { useAtom } from "jotai";
import { atomWithStorage } from "jotai/utils";
import { useEffect } from "react";

const darkModeAtom = atomWithStorage(
  "darkMode",
  false,
  window.localStorage,
  { getOnInit: true }
);

function App() {
  const [darkMode, setDarkMode] = useAtom(darkModeAtom);

  useEffect(() => {
    if (darkMode === false) {
      alert("dark");
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      <h1>Welcome to {darkMode ? "dark" : "light"} mode!</h1>
      <button onClick={() => setDarkMode(!darkMode)}>toggle theme</button>
    </>
  );
}

export default App;
```
Despite localStorage being set to true for darkMode, the initial render incorrectly triggers the alert("dark"), indicating that the atom is initialized with the default value instead of the stored value.

Comparison with Direct window.localStorage Usage
When I replaced atomWithStorage with a direct call to window.localStorage within my component, it correctly fetched and reflected the stored value from the beginning, as shown below:

```jsx
import "./App.css";
import { useAtom } from "jotai";
import { atomWithStorage } from "jotai/utils";
import { useEffect } from "react";

const darkModeAtom = atomWithStorage("darkMode", false);

function App() {
  const [, setDarkMode] = useAtom(darkModeAtom);

  const darkMode = window.localStorage.getItem("darkMode") === "true";

  useEffect(() => {
    if (darkMode === false) {
      alert("dark");
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      <h1>Welcome to {darkMode ? "dark" : "light"} mode!</h1>
      <button onClick={() => setDarkMode(!darkMode)}>toggle theme</button>
    </>
  );
}
```

This behavior confirms that using window.localStorage directly retrieves the expected value on initial render, contrasting with the behavior observed when using atomWithStorage.

Expected Behavior
When getOnInit is set to true, atomWithStorage should immediately use the value from localStorage upon initialization, ensuring the atom's state accurately reflects the stored value.

Actual Behavior
The atom initializes with the default value rather than the stored value from localStorage, even when getOnInit is set to true, which contrasts with the behavior when using window.localStorage directly.

This issue seems to suggest a potential bug or unexpected behavior in the handling of the getOnInit option within atomWithStorage. I would appreciate any insights or solutions to address this behavior.

**Comment by dai-shi** [comment, marked as answer]
Please try this:
```ts
const darkModeAtom = atomWithStorage(
  "darkMode",
  false,
  undefined,
  { getOnInit: true }
);
```

**Reply by chiwon1** [reply]
it solved. thanks!
